lib/spack/spack/cmd/install.py

```diff
@@ -393,6 +393,19 @@ def get_tests(specs):
     if args.deprecated:
         spack.config.set('config:deprecated', True, scope='command_line')
 
+    # concrete hashes in the env?
+    if args.spec and ev.active_environment() and args.only_concrete:
+        env = ev.active_environment()
+        hashes = [s.lstrip('/') for s in args.spec]
+        build = []
+        for (_, spec) in env.concretized_specs():
+            for s in spec.traverse(root=True):
+                for hash in hashes:
+                    if s.dag_hash().startswith(hash):
+                        build.append(s)
+        env.install_specs(specs=build)
+        return
```

Comment:
`args.spec` should be parsed -- move this down below `parse_specs` and use `specs`

Reply:
Agreed, this is a temporary hack, parsing queries the db and errors when the hash is not installed

Reply:
are you okay with https://github.com/spack/spack/pull/30276? first check the env, then the db? maybe it's not using the proper hash or whatever, but that should be fixed with your full hash everywhere pr.